python/lsst/jointcal/plot_chi2.py

```diff
@@ -0,0 +1,262 @@
+        self.ndof = np.array(self.ndof, dtype=np.int)
+        self.reduced = np.array(self.reduced, dtype=np.float64)
+        self.init_count = self._find_init()
+
+    def _find_init(self):
+        """Return the index of the first "fit step", after initialization.
+
+        NOTE
+        ----
+        There are never more than ~25 items in the list, so search optimization
+        is not worth the trouble.
+        """
+        # Logs pre-DM-25779
+        if "Fit prepared" in self.kind:
+            return self.kind.index("Fit prepared") + 1
+        # Logs post-DM-25779
+        elif "Fit iteration 0" in self.kind:
+            return self.kind.index("Fit iteration 0")
+        else:
+            raise RuntimeError(f"Cannot find end of initialization sequence in {self.kind}")
+
+
+class LogParser:
+    """Parse a jointcal logfile to extract chi2 values and plot them.
+
+    Call the instance with the path to a file to check it for anamolous chi2
+    and output plots to your current directory.
+
+    Parameters
+    ----------
+    plot : `bool`
+        Make plots for each file (saved to the current working directory)?
+    verbose : `bool`
+        Print updates during processing.
+    """
+    def __init__(self, plot=True, verbose=True):
+        chi2_re = "jointcal INFO: (?P<kind>.+) chi2/ndof : (?P<chi2>.+)/(?P<ndof>.+)=(?P<reduced_chi2>.+)"
+        self.matcher = re.compile(chi2_re)
+        self.plot = plot
+        self.verbose = verbose
+
+        # Reuse the Figure to speed up plotting and save memory.
+        self.fig = plt.figure(figsize=(15, 8))
+
+    def __call__(self, logfile):
+        """Parse logfile to extract chi2 values and generate and save plots.
+
+        The plot output is written to the current directory, with the name
+        derived from the basename of ``logfile``.
+
+        Parameters
+        ----------
+        logfile : `str`
+            The filename of the jointcal log to process.
+        """
+        title = os.path.basename(logfile)
+        if self.verbose:
+            print("Processing:", title)
+
+        astrometry = self._extract_chi2(logfile, "astrometry")
+        increased = self._find_chi2_increase(astrometry, title)
+        photometry = self._extract_chi2(logfile, "photometry")
+        increased |= self._find_chi2_increase(photometry, title)
+
+        if increased or self.plot:
+            self._plot(astrometry, photometry, title)
+            plt.savefig(f"{os.path.splitext(title)[0]}.png", bbox_inches="tight")
+
+    def _get_section_start_end(self, section):
+        """Return the start and end strings for this log section."""
+        if section.lower() == "astrometry":
+            return "Starting astrometric fitting...", "Updating WCS for visit:"
+        elif section.lower() == "photometry":
+            return "Starting photometric fitting...", "Updating PhotoCalib for visit:"
+        else:
+            raise RuntimeError(f"Do not understand log section type {section}.")
```

Comment:
Please add "Should be one of 'astrometry' or 'photometry'".

Reply:
I changed this to use a dictionary instead of a method call, so now it's a `KeyError`.